railties/lib/rails/application.rb

```diff
@@ -477,7 +477,21 @@ def secret_key_base
         config.secret_key_base ||= generate_local_secret
       else
         validate_secret_key_base(
-          ENV["SECRET_KEY_BASE"] || credentials.secret_key_base || secrets.secret_key_base
+          ENV["SECRET_KEY_BASE"] || credentials.secret_key_base || begin
+            secret_skb = secrets_secret_key_base
+
+            if secret_skb.equal?(config.secret_key_base)
```

Comment:
I'm not sure this is the correct location to check if `secrets.secret_key_base` equals `config.secret_key_base`.
I think a better location is where we do the assignment of `secrets.secret_key_base`:
https://github.com/rails/rails/blob/1abd331ff8b03a2f7325d3ba93a055ca7ad858b4/railties/lib/rails/application.rb#L451

Maybe change that to the following:
```ruby
  if secrets.secret_key_base.equal?(config.secret_key_base)
     Rails.deprecator.warn(<<~MSG.squish)
                Your `secret_key_base is configured in `Rails.application.secrets`,
                which is deprecated in favor of `Rails.application.credentials` and
                will be removed in Rails 7.2.
              MSG
            
  else
     secrets.secret_key_base ||= config.secret_key_base
  end
```

I think it also might make sense to move the updated deprecation in the memoized code block.

Reply:
Interesting, I think you're proposing something like this?

```diff
diff --git a/railties/lib/rails/application.rb b/railties/lib/rails/application.rb
index c3b0302a2c..50047507b9 100644
--- a/railties/lib/rails/application.rb
+++ b/railties/lib/rails/application.rb
@@ -448,7 +448,11 @@ def secrets
         secrets.merge! Rails::Secrets.parse(files, env: Rails.env)

         # Fallback to config.secret_key_base if secrets.secret_key_base isn't set
-        secrets.secret_key_base ||= config.secret_key_base
+        if secrets.secret_key_base
+          Rails.deprecator.warn("skb set in secrets")
+        else
+          secrets.secret_key_base = config.secret_key_base
+        end

         secrets
       end
@@ -477,7 +481,7 @@ def secret_key_base
         config.secret_key_base ||= generate_local_secret
       else
         validate_secret_key_base(
-          ENV["SECRET_KEY_BASE"] || credentials.secret_key_base || secrets.secret_key_base
+          ENV["SECRET_KEY_BASE"] || credentials.secret_key_base || secrets_secret_key_base
         )
       end
     end
```

This does pass the test I added, but I think this won't show the new deprecation warning if the only use of `secrets` is coming from `secret_key_base` (so I may need to add a second test for that). Alternatively if `secrets_secret_key_base` is kept as `secrets.secret_key_base` then the user will see both deprecation warnings which is part of what I want to avoid with this patch.

Reply:
Good point!
So we have two different deprecation warnings?

1. Using `config.secrets` directly in the app should use the existing deprecation warning as the user should be able to fix these.
2. Rails should output a different warning when the `secret_key_base` is still set/used in the `secrets`?

Maybe we can move the warnings to `secrets_secret_key_base`?
This would warn in development as well (although twice in the current implementation).

```ruby
      def secrets_secret_key_base
        secret_key_base = Rails.deprecator.silence do
          secrets.secret_key_base
        end
        if secret_key_base
          Rails.deprecator.warn("skb set in secrets")
          secret_key_base
        end
      end
```